### PR TITLE
Implement KL divergence for truncated normal distribution

### PR DESCRIPTION
**Pull request by breakds**
# Motivation

When KL divergence is needed in computing loss and when the projection network uses `TruncatedNormal`, an implementation of KL Divergence for `TruncatedNormal` is needed. We currently do not have it, and therefore we need to have it implemented for those cases (e.g. PPG Auxiliary Loss).

# Solution

Computed the KL Divergence closed form solution by hand and implemented it (turns out that it is **much harder** than I thought to get it right ...).

# Testing

Added an unit test to compare the solution with a numerical solution, and the result are consistent.

### CODE REVIEW

**Comment by breakds** [comment]
As requested, you can find [the deduction](https://files.breakds.org/wl/?id=T5RbjCi4tUmYLvYEEZPZgrhTZIGJZsc4) here.

File: alf/utils/distributions.py [hunk 1]
```diff
@@ -267,6 +267,54 @@ def __init__(self, loc, scale, lower_bound, upper_bound):
...
+
+    It is closed form w.r.t. torch.erf.
+
+    """
+    assert torch.all(
+        torch.logical_and(
+            torch.isclose(p.lower_bound, q.lower_bound),
+            torch.isclose(p.upper_bound, q.upper_bound)))
+
+    delta = (p.loc - q.loc)
```
Comment (emailweixu) [inline alf/utils/distributions.py]: () is not necessary

File: alf/utils/distributions.py [hunk 2]
```diff
@@ -267,6 +267,54 @@ def __init__(self, loc, scale, lower_bound, upper_bound):
...
+    delta2 = delta * delta
```
Comment (emailweixu) [inline alf/utils/distributions.py]: delta ** 2 might be actually faster because of special handling of **2 by pytorch (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/PowKernel.cu)

File: alf/utils/distributions.py [hunk 3]
```diff
@@ -267,6 +267,54 @@ def __init__(self, loc, scale, lower_bound, upper_bound):
...
+    # original gaussian (but both p and q are not normalized by area_p
+    # and area_q respectively). This will be handled below.
+    before_normalization = upper - lower
+
+    # Pad area_p as it is positive will only be served as denominator
+    area_p = p._cdf_ub - p._cdf_lb + 1e-30
+    area_q = q._cdf_ub - q._cdf_lb
+
+    return (torch.log(area_q / area_p) + before_normalization / area_p).sum(
+        dim=-1)
```
Comment (emailweixu) [inline alf/utils/distributions.py]: `sum(dim=list(range(-len(self._event_shape), 0)))`

File: alf/utils/distributions.py (hunk 1 above)
Comment (breakds) [reply alf/utils/distributions.py]: Removed.

File: alf/utils/distributions.py (hunk 2 above)
Comment (breakds) [reply alf/utils/distributions.py]: Good to learn. Updated all square computation to `** 2`.

File: alf/utils/distributions.py (hunk 3 above)
Comment (breakds) [reply alf/utils/distributions.py]: Good catch, thanks!

**Review by emailweixu — DISMISSED** [review]

**Comment by breakds** [comment]
Rebased `pytorch`.

**Review by emailweixu — APPROVED** [review]
